File: Solutions/2874. Maximum Value of an Ordered Triplet II/2874.py
class Solution:
    def maximumTripletValue(self, nums: List[int]) -> int:
        max_suffix = [0] * (len(nums) + 1)
        for i in range(len(nums)-1, 1, -1):
            max_suffix[i] = max(max_suffix[i+1], nums[i])
        i = nums[0]
        ans = 0
        for l in range(1, len(nums)-1):
            i = max(i, nums[l-1])
            k = max_suffix[l+1]
            j = nums[l]
            ans = max(ans, (i-j)*k)
        return ans

File: Solutions/2874. Maximum Value of an Ordered Triplet II/2874.go
func maximumTripletValue(nums []int) int64 {
    n := len(nums)
    max_suffix := make([]int, n+1)
    for i := n-1;  i > 0; i-- {
        max_suffix[i] = max(max_suffix[i+1], nums[i])
    }
    ans := 0
    i := nums[0]
    var j, k int
    for l := 1; l < n; l++{
        i = max(i, nums[l-1])
        j = nums[l]
        k = max_suffix[l+1]
        ans = max(ans, (i-j)*k)
    }
    return int64(ans)
}

